source/agora/common/crypto/Schnorr.d

```diff
@@ -0,0 +1,360 @@
+
+module agora.common.crypto.Schnorr;
+
+import agora.common.Data;
+import agora.common.Hash;
+
+import geod24.bitblob;
+import libsodium;
+
+/// Single signature example
+unittest
+{
+    Pair kp = Pair.random();
+    auto signature = sign(kp, "Hello world");
+    assert(verify(kp.V, signature, "Hello world"));
+}
+
+/// Multi-signature example
+unittest
+{
+    // Setup
+    static immutable string secret = "BOSAGORA for the win";
+    Pair kp1 = Pair.random();
+    Pair kp2 = Pair.random();
+    Pair R1 = Pair.random();
+    Pair R2 = Pair.random();
+    Point R = R1.V + R2.V;
+    Point X = kp1.V + kp2.V;
+
+    auto sig1 = sign(kp1.v, X, R, R1.v, secret);
+    auto sig2 = sign(kp2.v, X, R, R2.v, secret);
+    auto sig3 = Signature(R, sig1.s + sig2.s);
+
+    // No one can verify any of those individually
+    assert(!verify(kp1.V, sig1, secret));
+    assert(!verify(kp1.V, sig2, secret));
+    assert(!verify(kp2.V, sig2, secret));
+    assert(!verify(kp2.V, sig1, secret));
+    assert(!verify(kp1.V, sig3, secret));
+    assert(!verify(kp2.V, sig3, secret));
+
+    // But multisig works
+    assert(verify(X, sig3, secret));
+}
+
+/// Introduced in libsodium 1.0.18 - Need a fix to Geod24/libsodiumd#4 to remove
+private extern(C) void crypto_core_ed25519_scalar_mul(ubyte* z, const(ubyte)* x, const(ubyte)* y) nothrow @nogc;
+
+/// A compressed scalar
+public struct Scalar
+{
+    ///
+    public this (T) (T param)
+        if (is(typeof(this.data = typeof(this.data)(param))))
+    {
+        this.data = typeof(this.data)(param);
+    }
+
+    /// Reduce the hash to a scalar
+    public this (Hash param) nothrow @nogc
+    {
+        crypto_core_ed25519_scalar_reduce(this.data[].ptr, param[].ptr);
+    }
+
+    /// Internal state
+    private BitBlob!(crypto_core_ed25519_SCALARBYTES * 8) data;
+
+    /// Operator overloads
+    public Scalar opBinary (string op)(const scope auto ref Scalar rhs)
+        const nothrow @nogc
+    {
+        Scalar result = void;
+        static if (op == "+")
+            crypto_core_ed25519_scalar_add(
+                result.data[].ptr, this.data[].ptr, rhs.data[].ptr);
+        else static if (op == "-")
+            crypto_core_ed25519_scalar_sub(
+                result.data[].ptr, this.data[].ptr, rhs.data[].ptr);
+        else static if (op == "*")
+            crypto_core_ed25519_scalar_mul(
+                result.data[].ptr, this.data[].ptr, rhs.data[].ptr);
+        else
+            static assert(0, "Operator " ~ op ~ " not implemented");
+        return result;
+    }
+
+    public Scalar opUnary (string s)()
+        const nothrow @nogc
+        if (s == "-")
+    {
+        Scalar result = void;
+        crypto_core_ed25519_scalar_complement(result.data[].ptr, this.data[].ptr);
+        return result;
+    }
+
+    /// Generate a random scalar
+    public static Scalar random () nothrow @nogc
+    {
+        Scalar ret = void;
```

Comment:
Is the purpose of using these void initializers security? or performance?

Reply:
Both really

Reply:
Although I haven't thoroughly reviewed the code so it might not matter.